### ISSUE TITLE
RFC: Promote CouchDB 2.x support above CouchDB 1.x

### ISSUE THREAD

**Issue by willholley** [pouchdb/pouchdb#6343]
The PouchDB test suite is currently broken against CouchDB master (2.x). Whilst investigating this, I noticed that some core behaviour is starting to diverge between 1.x and 2.x in small ways - bugs getting fixed so that certain errors return more meaningful errors/status codes, etc - which makes testing against both versions increasingly awkward.

I propose that:
 * Where relevant, we adjust PouchDB to match the API (status codes, errors) of the latest CouchDB release.
 * After this round of fixes, we start requiring that the couch-master build in Travis passes for new PRs
 * We skip tests against CouchDB 1.x where the behaviour is no longer matched by PouchDB.

Just to be clear, I'm not proposing that PouchDB drops support for CouchDB 1.x - just that 2.x/master should be the primary target for API compatibility.

**Comment by garrensmith** [comment]
I agree with this. CouchDB 2.x is the future, I would rather we support it as much as possible and factor in a little divergence from 1.x

**Comment by daleharvey** [comment]
Agree 100%, of course we need to get master green, but once it is fully agreed its passing tests being blocking and to avoid having divergent tests for 1 + 2 and focussing on 2

**Comment by nolanlawson** [comment]
I agree, but I would rather have tests passing for both CouchDB 1 and 2 since there are still lots of people depending on 1.6. :smiley:

My hunch is that many of the failed tests revolve around stuff that isn't really important to test precisely, e.g. "when I get a 403, does the error message looks _exactly_ like this string." In those cases we should just make the tests more generic, check that there's a string, etc.

We also have `testUtils.isCouchMaster()` which should make it fairly easy to have test for one versus the other.

So basically I would support points 1 and 2 but not 3 because there is still some fairly important stuff that can only be tested in CouchDB 1 or PouchDB Server (e.g. temporary views, which we can consider fully deprecated once pouchdb-mapreduce is moved out of core).

**Comment by willholley** [comment]
@nolanlawson agree that we shouldn't skip tests against CouchDB 1.x except where there is conflicting behaviour (PouchDB and CouchDB.latest behave one way, CouchDB.old behaves another). 

A small but concrete example is that invalid input to `_bulk_docs` returns a 500 status code in CouchDB 1.x/2.0 but 4XX in master. I'd favour changing PouchDB and the test to track CouchDB master and just skip it for 1.x.

[COUCHDB-2386](https://issues.apache.org/jira/browse/COUCHDB-2386) is maybe an example of a more significant change. As a minimum, we need to change the tests around updating deleted docs, but it's questionable whether PouchDB should match the new CouchDB behaviour, particularly when it's unclear whether the change is intended or not.

**Comment by nolanlawson** [comment]
Fair enough, I'm happy as long as we're still running tests against 1.6 somewhere. I expect it'll still remain popular for at least the next few years.

COUCHDB-2386 seems like a true bug to me, but we can skip it until CouchDB fixes it. It looks super edge-casey to me.

**Comment by jahtalab** [comment]
What is the status of supporting CouchDB 2.x?
Is there any place to follow on the remaining issues?

**Comment by willholley** [comment]
CouchDB 2 is fully supported today - this ticket is just about switching
out testing configuration so that 2.x becomes the default target rather
than a special case.

**Comment by jahtalab** [comment]
Thanks @willholley!
